# **To Do 1**

**1. Read and Observe the Dataset.**

In [31]:
import pandas as pd
df = pd.read_csv('drive/MyDrive/Dataset/student.csv');
df

Math  Reading  Writing
0      48       68       63
1      62       81       72
2      79       80       78
3      76       83       79
4      59       64       62
..    ...      ...      ...
995    72       74       70
996    73       86       90
997    89       87       94
998    83       82       78
999    66       66       72

[1000 rows x 3 columns]

**2. Print top(5) and bottom(5) of the dataset {Hint: pd.head and pd.tail}.**

In [32]:
import pandas as pd
df = pd.read_csv('drive/MyDrive/Dataset/student.csv');
pd = df.head(5)
pd

Math  Reading  Writing
0    48       68       63
1    62       81       72
2    79       80       78
3    76       83       79
4    59       64       62

In [33]:
import pandas as pd
df = pd.read_csv('drive/MyDrive/Dataset/student.csv');
pd = df.tail(5)
pd

Math  Reading  Writing
995    72       74       70
996    73       86       90
997    89       87       94
998    83       82       78
999    66       66       72

**3. Print the Information of Datasets. {Hint: pd.info}.**

In [34]:
pd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 995 to 999
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype
---  ------   --------------  -----
 0   Math     5 non-null      int64
 1   Reading  5 non-null      int64
 2   Writing  5 non-null      int64
dtypes: int64(3)
memory usage: 252.0 bytes


**4. Gather the Descriptive info about the Dataset. {Hint: pd.describe}**

In [35]:
pd.describe()

Math    Reading    Writing
count   5.0000   5.000000   5.000000
mean   76.6000  79.000000  80.800000
std     9.2358   8.888194  10.733126
min    66.0000  66.000000  70.000000
25%    72.0000  74.000000  72.000000
50%    73.0000  82.000000  78.000000
75%    83.0000  86.000000  90.000000
max    89.0000  87.000000  94.000000

**5. Split your data into Feature (X) and Label (Y).**

In [36]:
X = df[['Math','Reading']]
Y = df['Writing']

# **To - Do - 2:**

**1. To make the task easier - let’s assume there is no bias or intercept.**

**2. Create the following matrices:**

Y = W^T.X

In [37]:
import numpy as np
import pandas as pd
df = pd.read_csv('drive/MyDrive/Dataset/student.csv');
X = df[['Math','Reading']].values.T
Y = df['Writing'].values
d = X.shape[0]
W = np.random.rand(d, 1)
Y_pred = np.dot(W.T, X)

print("Feature Matrix X:")
print(X)
print("Shape of X:", X.shape)

print("\nWeight Vector W:")
print(W)
print("Shape of W:", W.shape)

print("\nActual Output Y:")
print(Y)
print("Shape of Y:", Y.shape)

print("\nPredicted Output Ŷ:")
print(Y_pred)
print("Shape of Ŷ:", Y_pred.shape)

Feature Matrix X:
[[48 62 79 ... 89 83 66]
 [68 81 80 ... 87 82 66]]
Shape of X: (2, 1000)

Weight Vector W:
[[0.79654299]
 [0.18343479]]
Shape of W: (2, 1)

Actual Output Y:
[ 63  72  78  79  62  85  83  41  80  77  64  90  45  77  70  46  76  44
  85  72  53  66  75  49  84  83  68  66  77  78  74  83  72  65  46  66
  50  79  68  46  86  70  61  53  72  75  50  77 100  81 100  87  78  48
  50  44  48  43  67  78  58  91  92  78  42  85  73  83  61  58  60  55
  48  62  68  59  62  48  74  63  80  79  73  79  45  67  89  77  81  88
  53  68  79  77  63  73  60  67 100  79  26  51  80  57  41  78  68  49
  76  41  71  77  89  86  55  80  56  74  85  80  73  74  86  56  53  44
  41  59  71  81  74  78  67  53  56  75  82  79  99  76  59  96  75  61
  56  88  65 100  79  55  61  83  74  59  54  47  82  74  59  74  84  59
  43  65  61  78  84  73  73  92  63  72  61  59  70  87  78  65  73  62
  69  55  73  63  67  86  78  85  83  80  60  90  56  70  55  80  82  60
  78  76  94  75  68  

# **• To - Do - 3:**

1. Split the dataset into training and test sets.
2. You can use an 80-20 or 70-30 split, with 80% (or 70%) of the data used for training and the rest
for testing.

In [38]:
import pandas as pd
import numpy as np
df = pd.read_csv('drive/MyDrive/Dataset/student.csv');
X = df[['Math', 'Reading']].values.T
Y = df['Writing'].values

def train_test_split_scratch(X, y, test_size=0.2, random_seed=42):
    np.random.seed(random_seed)
    n = X.shape[1]

    indices = np.arange(n)
    np.random.shuffle(indices)

    test_size_count = int(n * test_size)

    test_indices = indices[:test_size_count]
    train_indices = indices[test_size_count:]

    X_train = X[:, train_indices]
    X_test = X[:, test_indices]
    y_train = y[train_indices]
    y_test = y[test_indices]

    return X_train, X_test, y_train, y_test

X_train, X_test, Y_train, Y_test = train_test_split_scratch(X, Y)

# Print shapes to verify
print("X_train shape:", X_train.shape)
print("Y_train shape:", Y_train.shape)
print("X_test shape:", X_test.shape)
print("Y_test shape:", Y_test.shape)


X_train shape: (2, 800)
Y_train shape: (800,)
X_test shape: (2, 200)
Y_test shape: (200,)


# **To - Do - 4:**

Feel free to build your own code or complete the following code:

In [39]:
def cost_function(X, Y, W):
  n = X.shape[1]
  Y_pred = np.dot(W.T, X)
  error = Y_pred - Y
  cost = (1 / (2 * n)) * np.sum(error ** 2)
  return cost


# **To - Do - 5:**

Make sure your code at To - Do - 4 passed the following test case:

In [40]:
import numpy as np

X_test = np.array([[1, 3, 5],
                   [2, 4, 6]])
Y_test = np.array([3, 7, 11])
W_test = np.array([[1],
                   [1]])

cost = cost_function(X_test, Y_test, W_test)

if cost == 0:
    print("Proceed Further")
else:
    print("Something went wrong: Reimplement the cost function")
    print("Cost function output:", cost)


Proceed Further


# **To - Do - 6:**

Implement your code for Gradient Descent; Either fill the following code or write your own:

In [41]:
import numpy as np

def gradient_descent(X, Y, W, alpha, iterations):
    m = X.shape[0]
    cost_history = []

    for _ in range(iterations):

        Y_pred = np.dot(X, W)

        error = Y_pred - Y

        dw = (1 / m) * np.dot(X.T, error)

        W = W - alpha * dw

        cost = (1 / (2 * m)) * np.sum(error ** 2)
        cost_history.append(cost)

    return W, cost_history



# **To - Do - 7:**

Make sure following Test Case is passe by your code from To - Do - 6 or your Gradient Descent
Implementation:

In [42]:
np.random.seed(0)

X = np.random.rand(100, 3)
Y = np.random.rand(100)
W = np.random.rand(3)

alpha = 0.01
iterations = 1000

final_params, cost_history = gradient_descent(X, Y, W, alpha, iterations)

print("Final Parameters:", final_params)
print("Final Cost:", cost_history[-1])


Final Parameters: [0.20551667 0.54295081 0.10388027]
Final Cost: 0.05435809373901896


# **To - Do - 8:**

Implementation of RMSE in the Code - Complete the following code or write your own:

In [43]:
import numpy as np
def rmse(Y, Y_pred):
    n = len(Y)
    error = Y - Y_pred
    rmse = np.sqrt((1 / n) * np.sum(error ** 2))

    return rmse


# **To - Do - 9 - Implementation in the Code:**

Complete the following code or write your own for r2 loss:

In [44]:
import numpy as np

def r2(Y, Y_pred):
    mean_y = np.mean(Y)
    ss_tot = np.sum((Y - mean_y) ** 2)
    ss_res = np.sum((Y - Y_pred) ** 2)
    r2 = 1 - (ss_res / ss_tot)

    return r2


# **• To - Do - 10:**

We will define a function that:

1. Loads the data and splits it into training and test sets.

2. Prepares the feature matrix (X) and target vector (Y).

3. Defines the weight matrix (W) and initializes the learning rate and number of iterations.

4. Calls the gradient descent function to learn the parameters.

5. Evaluates the model using RMSE and R2

Re-write the following code or Write your own:

In [47]:
def main():

    data = pd.read_csv('drive/MyDrive/Dataset/student.csv');

    X = data[['Math', 'Reading']].values
    Y = data['Writing'].values

    np.random.seed(42)
    indices = np.arange(len(X))
    np.random.shuffle(indices)
    split = int(0.8 * len(X))

    X_train = X[indices[:split]]
    X_test  = X[indices[split:]]
    Y_train = Y[indices[:split]]
    Y_test  = Y[indices[split:]]

    W = np.zeros(X_train.shape[1])
    alpha = 0.00001
    iterations = 1000

    W_opt, cost_history = gradient_descent(X, Y, W, alpha, iterations)

    Y_pred = np.dot(X_test, W_opt)

    print("Final Weights:", W_opt)
    print("RMSE:", rmse(Y_test, Y_pred))
    print("R2 Score:", r2(Y_test, Y_pred))


if __name__ == "__main__":
    main()

Final Weights: [0.35019408 0.64494005]
RMSE: 5.240404248780206
R2 Score: 0.8697446485664456


# **To - Do - 11 - Present your finding:**

1. Did your Model Overfitt, Underfitts, or performance is acceptable.
2. Experiment with different value of learning rate, making it higher and lower, observe the result.

In [52]:
def main():
    # Load dataset
    data = pd.read_csv('drive/MyDrive/Dataset/student.csv')

    X = data[['Math', 'Reading']].values
    Y = data['Writing'].values

    # Shuffle and split
    np.random.seed(42)
    indices = np.arange(len(X))
    np.random.shuffle(indices)
    split = int(0.8 * len(X))

    X_train = X[indices[:split]]
    X_test  = X[indices[split:]]
    Y_train = Y[indices[:split]]
    Y_test  = Y[indices[split:]]

    # Initial weights and iterations
    W = np.zeros(X_train.shape[1])
    iterations = 1000

    # Experiment with different learning rates
    alphas = [0.000001, 0.00001, 0.0001]

    for alpha in alphas:
        W_opt, cost_history = gradient_descent(X_train, Y_train, W, alpha, iterations)

        Y_train_pred = np.dot(X_train, W_opt)
        Y_test_pred  = np.dot(X_test, W_opt)

        print(f"\nLearning Rate: {alpha}")
        print("Final Weights:", W_opt)
        print("Train RMSE:", rmse(Y_train, Y_train_pred))
        print("Test RMSE:", rmse(Y_test, Y_test_pred))
        print("Train R²:", r2(Y_train, Y_train_pred))
        print("Test R²:", r2(Y_test, Y_test_pred))


if __name__ == "__main__":
    main()


Learning Rate: 1e-06
Final Weights: [0.47500971 0.52321817]
Train RMSE: 5.753099907093657
Test RMSE: 5.912873961371951
Train R²: 0.8605053065080729
Test R²: 0.8341699475180014

Learning Rate: 1e-05
Final Weights: [0.3533225  0.64122757]
Train RMSE: 5.161226707460374
Test RMSE: 5.258385317376058
Train R²: 0.8877310329425561
Test R²: 0.8688492410749991

Learning Rate: 0.0001
Final Weights: [0.09473512 0.89188962]
Train RMSE: 4.575335135099187
Test RMSE: 4.61158464525486
Train R²: 0.9117733728428985
Test R²: 0.8991289912947937


Based on the gradient descent experiments with different learning rates, the model shows varying performance. With a very low learning rate of 0.000001, the model underfits slightly, as indicated by a relatively higher Train RMSE of 5.75 and Test RMSE of 5.91, with R² scores of 0.86 and 0.83 respectively. Increasing the learning rate to 0.00001 improves performance, reducing Train and Test RMSE to 5.16 and 5.26, and increasing R² scores to 0.89 (train) and 0.87 (test), indicating better learning and generalization. A further increase to 0.0001 results in the best performance among the tested rates, with Train RMSE of 4.58, Test RMSE of 4.61, and R² scores of 0.91 (train) and 0.90 (test). Overall, the model does not overfit, as train and test errors are close, and the performance is acceptable. The optimal learning rate for this dataset appears to be 0.0001, which balances fast convergence and stable, accurate predictions.